contracts/ScheduledPaymentModule.sol

```diff
@@ -78,6 +98,103 @@ contract ScheduledPaymentModule is Module {
         emit ScheduledPaymentCancelled(spHash);
     }
 
+    function executeScheduledPayment(
+        address token,
+        uint256 amount,
+        address payee,
+        uint256 executionGas,
+        uint256 maxGasPrice,
+        address gasToken,
+        uint256 _nonce,
+        uint256 payAt,
+        uint256 gasPrice
+    ) external onlyCrank {
+        uint256 startGas = gasleft();
+        bytes32 spHash = createSpHash(
+            token,
+            amount,
+            payee,
+            executionGas,
+            maxGasPrice,
+            gasToken,
+            _nonce,
+            payAt
+        );
+        if (!spHashes.contains(spHash)) revert UnknownHash(spHash);
+
+        // 1 minute is buffer to protect against miners gaming block time
+        // The recommended time for POW consensus finality is 1 minute
+        if (block.timestamp < payAt.add(1 minutes))
+            revert InvalidPeriod(spHash);
+        if (gasPrice > maxGasPrice) revert ExceedMaxGasPrice(spHash);
+
+        if (
+            !executePayment(
+                token,
+                amount,
+                payee,
+                executionGas,
+                gasPrice,
+                gasToken
+            )
+        ) revert PaymentExecutionFailed(spHash);
+        if (startGas - gasleft() + 2500 + 500 > executionGas)
+            revert OutOfGas(spHash);
+
+        spHashes.remove(spHash);
+        emit ScheduledPaymentExecuted(spHash);
+    }
+
+    function estimateExecutionGas(
+        address token,
+        uint256 amount,
+        address payee,
+        uint256 executionGas,
+        uint256 maxGasPrice,
+        address gasToken,
+        uint256 _nonce,
+        uint256 payAt,
+        uint256 gasPrice
+    ) external returns (uint256) {
```

Comment:
would it be possible for this function to just wrap `executeScheduledPayment`, like:

```
function estimateExecutionGas(
        address token,
        uint256 amount,
        address payee,
        uint256 executionGas,
        uint256 maxGasPrice,
        address gasToken,
        uint256 _nonce,
        uint256 payAt,
        uint256 gasPrice
    ) external returns (uint256) {
      uint256 startGas = gasleft();
      executeScheduledPayment(......);
      uint256 requiredGas = startGas - gasleft() + 2500 + 500;
      revert GasEstimation(requiredGas);
    }
```
This way we don't have to worry about the implementation of the gas measurement getting out of sync with teh actual call itself. also it sames on duplicated code

Reply:
The problem is that the `executeScheduledPayment` function has multiple validation processes. One is to check if the hash is already scheduled or not. I think the `estimateExecutionGas` function might be called before the user schedules the payment hash. This will cause the process to revert the UnknownHash error instead of the GasEstimation error as expected. Actually, we can use `executioGas` parameter as a flag to skip the validation process, if `executionGas` is equal to 0 then the validation process will be skipped, but in my opinion, it is too risky because there may be a bug or error that causes Crank to incorrectly call the `executeScheduledPayment` function to perform scheduled payments are not for gas estimates but misplaced gas estimates equal to 0.

Reply:
Can’t we just make a private _executeScheduledPayment() that doesn’t include the require?,

then executeTransaction() can perform the require() checks and then call the private _executeTransaction to do the actual logic. The estimateGas could call the same private _executeTransaction. The estimateGas will not be able to count the require checks as part of the gas, but that would be considered to be part of the baseGas.

Reply:
This private function also not include the hashing process to generates spHash? Because the require checks need it to do validations. Or we do the require checks after call the private function and the private function return the status of execution and spHash?

Reply:
sure we can leave that out too--and that will also need to be counted as part of the baseGas

Reply:
updated